File: frontend/src/app/core/services/chart.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';

export interface SearchResult {
  symbol: string;
  name: string;
  exchange: string;
  type: string;
}

export interface StockQuote {
  symbol: string;
  name: string;
  price: number;
  previousClose: number;
  change: number;
  pctChange: number;
  open: number;
  dayHigh: number;
  dayLow: number;
  volume: number;
  marketCap: number;
  fiftyTwoWeekHigh: number;
  fiftyTwoWeekLow: number;
  exchange: string;
  currency: string;
  logo: string;
}

export interface OHLCVData {
  time: number | string;
  date?: string;
  open: number;
  high: number;
  low: number;
  close: number;
  volume: number;
}

export interface HistoryResponse {
  symbol: string;
  history: OHLCVData[];
  period: string;
  interval: string;
  count: number;
}

export interface Fundamental {
  symbol: string;
  name: string;
  sector: string;
  industry: string;
  marketCap: number;
  pe: number;
  forwardPe: number;
  eps: number;
  bookValue: number;
  dividendYield: number;
  roe: number;
  debtToEquity: number;
  fiftyTwoWeekHigh: number;
  fiftyTwoWeekLow: number;
  fiftyDayAvg: number;
  twoHundredDayAvg: number;
  beta: number;
  totalRevenue: number;
  revenueGrowth: number;
  profitMargin: number;
  operatingMargin: number;
  website: string;
  description: string;
}

export interface NewsArticle {
  title: string;
  publisher: string;
  link: string;
  publishedAt: number;
  thumbnail: string;
}

export interface OptionEntry {
  strike: number;
  lastPrice: number;
  change: number;
  pctChange: number;
  volume: number;
  openInterest: number;
  impliedVolatility: number;
  inTheMoney: boolean;
}

export interface OptionChain {
  symbol: string;
  expiryDate: string;
  calls: OptionEntry[];
  puts: OptionEntry[];
}

@Injectable({ providedIn: 'root' })
export class ChartService {
  private apiUrl = `${import.meta.env.NG_APP_BACKEND}/api/charts`;

  constructor(private http: HttpClient) { }

  searchTickers(query: string): Observable<{ results: SearchResult[] }> {
    return this.http.get<{ results: SearchResult[] }>(`${this.apiUrl}/search`, {
      params: { q: query }
    });
  }

  getQuote(symbol: string): Observable<StockQuote> {
    return this.http.get<StockQuote>(`${this.apiUrl}/quote`, {
      params: { symbol }
    });
  }

  getHistory(symbol: string, period = '1y', interval = '1d'): Observable<HistoryResponse> {
    return this.http.get<HistoryResponse>(`${this.apiUrl}/history`, {
      params: { symbol, period, interval }
    });
  }

  getIntraday(symbol: string, interval = '1m'): Observable<HistoryResponse> {
    return this.http.get<HistoryResponse>(`${this.apiUrl}/intraday`, {
      params: { symbol, interval }
    });
  }

  getFundamentals(symbol: string): Observable<Fundamental> {
    return this.http.get<Fundamental>(`${this.apiUrl}/fundamentals`, {
      params: { symbol }
    });
  }

  getFinancials(symbol: string, statement = 'income', quarterly = false): Observable<any> {
    return this.http.get(`${this.apiUrl}/financials`, {
      params: { symbol, statement, quarterly: quarterly.toString() }
    });
  }

  getNews(symbol: string): Observable<{ symbol: string; articles: NewsArticle[] }> {
    return this.http.get<{ symbol: string; articles: NewsArticle[] }>(`${this.apiUrl}/news`, {
      params: { symbol }
    });
  }

  getOptionDates(symbol: string): Observable<{ symbol: string; dates: string[] }> {
    return this.http.get<{ symbol: string; dates: string[] }>(`${this.apiUrl}/options/dates`, {
      params: { symbol }
    });
  }

  getOptionChain(symbol: string, date: string): Observable<OptionChain> {
    return this.http.get<OptionChain>(`${this.apiUrl}/options/chain`, {
      params: { symbol, date }
    });
  }
}

File: mobile/lib/core/services/chart_service.dart
import 'dart:convert';
import 'api_service.dart';
import '../../config/api_config.dart';

class ChartService {
  Future<Map<String, dynamic>> searchTickers(String query) async {
    final response = await ApiService.get(
      '${ApiConfig.apiUrl}/charts/search?q=$query',
    );
    if (response.statusCode == 200) {
      return json.decode(response.body);
    }
    throw Exception('Failed to search tickers');
  }

  Future<Map<String, dynamic>> getQuote(String symbol) async {
    final response = await ApiService.get(
      '${ApiConfig.apiUrl}/charts/quote?symbol=$symbol',
    );
    if (response.statusCode == 200) {
      return json.decode(response.body);
    }
    throw Exception('Failed to get quote');
  }

  Future<Map<String, dynamic>> getHistory(
    String symbol,
    String period,
    String interval,
  ) async {
    final response = await ApiService.get(
      '${ApiConfig.apiUrl}/charts/history?symbol=$symbol&period=$period&interval=$interval',
    );
    if (response.statusCode == 200) {
      return json.decode(response.body);
    }
    throw Exception('Failed to load chart data');
  }

  Future<Map<String, dynamic>> getFundamentals(String symbol) async {
    final response = await ApiService.get(
      '${ApiConfig.apiUrl}/charts/fundamentals?symbol=$symbol',
    );
    if (response.statusCode == 200) {
      return json.decode(response.body);
    }
    throw Exception('Failed to load fundamentals');
  }

  Future<Map<String, dynamic>> getFinancials(
    String symbol, {
    String statement = 'results',
  }) async {
    final response = await ApiService.get(
      '${ApiConfig.apiUrl}/charts/financials?symbol=$symbol&statement=$statement',
    );
    if (response.statusCode == 200) {
      return json.decode(response.body);
    }
    throw Exception('Failed to load financials');
  }

  Future<Map<String, dynamic>> getNews(String symbol) async {
    final response = await ApiService.get(
      '${ApiConfig.apiUrl}/charts/news?symbol=$symbol',
    );
    if (response.statusCode == 200) {
      return json.decode(response.body);
    }
    throw Exception('Failed to load news');
  }

  Future<Map<String, dynamic>> getOptionDates(String symbol) async {
    final response = await ApiService.get(
      '${ApiConfig.apiUrl}/charts/options/dates?symbol=$symbol',
    );
    if (response.statusCode == 200) {
      return json.decode(response.body);
    }
    throw Exception('Failed to load option dates');
  }

  Future<Map<String, dynamic>> getOptionChain(
    String symbol,
    String date,
  ) async {
    final response = await ApiService.get(
      '${ApiConfig.apiUrl}/charts/options/chain?symbol=$symbol&date=$date',
    );
    if (response.statusCode == 200) {
      return json.decode(response.body);
    }
    throw Exception('Failed to load option chain');
  }
}
